zeppelin-interpreter/src/main/java/org/apache/zeppelin/interpreter/InterpreterGroup.java

```diff
@@ -168,13 +172,7 @@ public void close(String sessionId) {
     List<Interpreter> intpForSession = this.get(sessionId);
     close(intpForSession);
 
-    if (remoteInterpreterProcess != null) {
-      remoteInterpreterProcess.dereference();
-      if (remoteInterpreterProcess.referenceCount() <= 0) {
-        remoteInterpreterProcess = null;
-        allInterpreterGroups.remove(id);
-      }
-    }
+    makeSureRemoteProcessTerminate();
```

Comment:
`remoteInterpreterProcess.dereference()` looked like a Array.pop() to me.

Also, I've checked for this use case; Have 1st interpreter say livy (without user impersonation) and other shell (with user impersonation which also means, interpreter will be instantiated Per User in isolated  process). Ran these two in two different notebooks, check what all process in running, then restart interpreter one-by-one for each user.
 
![check process](https://cloud.githubusercontent.com/assets/674497/23091306/272fface-f5d9-11e6-92b8-535a7b714189.gif)


Can you list an example where this could fail, I'll try and fix.